clang/lib/AST/PreorderAST.cpp

```diff
@@ -56,11 +62,58 @@ void PreorderAST::Create(Expr *E, Node *Parent) {
   E = Lex.IgnoreValuePreservingOperations(Ctx, E->IgnoreParens());
 
   if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
-    auto *N = new BinaryNode(BO->getOpcode(), Parent);
+    BinaryOperator::Opcode BinOp = BO->getOpcode();
+    Expr *LHS = BO->getLHS();
+    Expr *RHS = BO->getRHS();
+
+    // We can convert (e1 - e2) to (e1 + -e2) if -e2 does not overflow.  One
```

Comment:
I think you should only put this into a normal form if -e2 is guaranteed to not overflow.   That means that if e2 is not an integer constant, it should not be put in a normal form.